Add check for improperly torn-down host aggregates

AnishReddyRavula opened the pull request:
When a reservation ends, the hosts reserved by it are supposed to be moved from the reservation's aggregate into the freepool aggregate.

Sometimes, this doesn't happen correctly, so there is additional code added to the host monitor to clean up these errors periodically.

AnishReddyRavula committed:
Add check for improperly torn-down host aggregates

When a reservation ends, the hosts reserved by it are supposed to be moved from the reservation's aggregate into the freepool aggregate.

Sometimes, this doesn't happen correctly, so there is additional code added to the host monitor to clean up these errors periodically.

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,39 @@ def poll_resource_failures(self):
 
+            aggregates = self.nova.aggregates.list()
+            if aggregates:
+                freepool = next((agg for agg in aggregates if agg.name == "freepool"), None)
```
I am concerned this might be slow, and I wonder if there is a better API for it. I see there is `get_details` in the sdk, but not sure what exact versions we are using. https://docs.openstack.org/python-novaclient/latest/reference/api/novaclient.v2.aggregates.html

Mark-Powers replied:
I guess it actually doesn't matter, since we are iterating through the list later in the function.

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,39 @@ def poll_resource_failures(self):
+                    for host in agg.hosts:
+                        try:
+                            # Remove the host from the reservation's aggregate
+                            agg.remove_host(host)
+                            # Return the host to the freepool
+                            freepool.add_host(host)
+                            recovered_hosts.append(host)
```
I am wondering if it is possible for this to fail, and a node to be "lost" without an aggregate as a result.

AnishReddyRavula replied:
I think the reason this would fail is if the host is not found, or if the freepool already has the host. 
I am not sure how to enforce transaction operation here - maybe retry the operation?

Mark-Powers reviewed (approved):
Looks good overall, and I think this is better than what we have. I'm not sure about how to safely move these between aggregates to ensure we don't have to do further clean up in case this monitor fails. @msherman64 any ideas?

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,39 @@ def poll_resource_failures(self):
+                    LOG.warning(
+                        f"Host aggregate for reservation {agg.name} failed to clean up."
+                    )
+                    for host in agg.hosts:
```
Can this be done in parallel?

msherman64 reviewed (changes requested):
Summarizing from our conversation earlier:

This approach lists hosts which are in aggregates, but won't catch hosts not in an aggregate (which can happen)

We should iterate over blazar hosts, and for each host, use the blazar allocation and reservation data to compute what aggregates it "should" be in.

Then, we can add some checks:
1. if a host is not reserved, it should be in the freepool
2. if a host is in more than one aggregate, it should be removed from whichever aggregate is not correct.

Two race conditions to be concerned with in this code:
1. if a lease starts, the on_lease_start code will remove the host from the freepool, and add it to the reservation aggregate. Therefore, we don't want to remove hosts from the freepool in this new code.
2. If a lease ends, the on_lease_end code will remove the host from its aggregate, and add it to the freepool. Therefore we dont' want to remove hosts from aggregates if that aggregate corresponds to a lease that is "good", meaning NOT in state "deleted" or "error" @Mark-Powers does this seem correct?

AnishReddyRavula committed:
Iterate hosts and move hosts to freepool based on reservation status

create a new db utils method to collect host's non pending reservation details

By iterating through hosts and getting the most recent non "pending" reservation on the host
if the reservation is active, then everythig is alright - the host must be in the right aggregate.
but if it is in deleted or error state, then we need to move the host to freepool

Mark-Powers commented on blazar/db/sqlalchemy/utils.py:
```diff
@@ -293,6 +294,49 @@ def get_reservation_allocations_by_device_ids(device_ids, start_date, end_date,
+        host_id (): Host id - primary key of ComputeHost table
+    """
+    session = get_session()
+    curr_date = datetime.now()
```
can you add some time (maybe ~3-5 minutes) to this to ensure a lease that is about to start won't start while the cleanup is running?

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,42 @@ def poll_resource_failures(self):
+                    # if the host is not in any aggregate and in non-active reservation
+                    # it should be moved to freepool
+                    freepool.add_host(host_uuid)
+                    recovered_hosts.append(host)
```
You shouldn't touch recovered_hosts here, unless I am missing something. `recovered_hosts` and `failed_hosts` are meant to be hosts that were marked as unreservable, or should be set as unreservable. Here, since you've fixed the problem, the hosts don't need to be updated.

Mark-Powers commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,42 @@ def poll_resource_failures(self):
+                # get the most recent reservation for the host_id and check if we need to move to freepool
+                reservation = db_utils.get_recent_non_pending_reservation_by_host_id(host.id)
+                # an aggregate not in 'active' or 'pending' reservation should be in freepool
+                if reservation.status == status.reservation.ACTIVE:
```
So essentially, a reservation only gets past this point if the reservation is deleted or in error? That seems like correct behavior.

Mark-Powers reviewed (approved):
Overall, I think this looks good and is correct. I am not sure how slow this will take to run, as the aggregate API can be slow. 

To test this, you should be able to create a lease, wait for it to end, and then recreate the aggregate with `openstack aggregate` commands and move a host back into it. Then I would expect this code to undo the commands you run.

AnishReddyRavula committed:
Addressing review comments - added current time by 5 minuts to skip leases about to start

removed adding to recovered hosts and failed hosts

Added test cases

AnishReddyRavula committed:
Better log messages for host cleanup

AnishReddyRavula committed:
db returns are dicts, modify code to use dicts instead for hosts and reservations

along with few code fixes

msherman64 commented on blazar/db/sqlalchemy/utils.py:
```diff
@@ -293,6 +294,43 @@ def get_reservation_allocations_by_device_ids(device_ids, start_date, end_date,
     return reservations
 
 
+def get_recent_non_pending_reservation_by_host_id(host_id):
```
Making sure I understand the method name (and purpose)

This is getting some parameters from the computhost, reservation, lease, and computehostreservation tables.
The results are restricted to those with the current computehost ID, and the lease isn't pending (meanning it has started or some action taken), AND the start date is in the past.

Finally we call "first()" to get only the values corresponding to the most recently started lease, rather than a list.

I was initally thrown off by the name, as it doesn't appear to return a `reservation`.
Perhaps something like:

get_most_recent_reservation_info_by_host_id? I'm not quite sure what to refer to the collection of data as, since it's not from a single table or class.

msherman64 commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,46 @@ def poll_resource_failures(self):
+                    LOG.info(f"{host['hypervisor_hostname']} is in a active reservation"
+                             f"{reservation['reservation_id']}")
+                    continue
+                # host not in 'active' or 'pending' reservation should be in freepool
```
this makes sense, but is a little bit confusing because we filter out "ACTIVE" reservations with the `continue`, and filter out `PENDING` reservations in the DB query. Would it be cleaner to filter out both active and pending reservations in the DB query?

AnishReddyRavula replied:
Actually no. We would like to know if the host is currently in a `active` reservations so we can skip it as "it is already  in the right freepool" otherwise it would be in errored state

msherman64 commented on blazar/plugins/oshosts/host_plugin.py:
```diff
@@ -973,6 +973,46 @@ def poll_resource_failures(self):
+                # host not in 'active' or 'pending' reservation should be in freepool
+                host_uuid = host['hypervisor_hostname']
+                curr_agg = host_to_agg_map.get(host['hypervisor_hostname'])
+                if curr_agg and (curr_agg.name == freepool.name):
```
Just convincing myself why we are only comparing to "freepool", rather than also to the target aggregate from the current reservationm.

`curr_agg` comes from `host_to_agg_map` which comes from `self.nova.aggregates.list()`
so, this is the value returned by nova?

this is in contrast to the value we get from the blazar DB, `models.ComputeHostReservation.aggregate_id`, which is the "target" aggregate, since that's only relevant during `active` or `pending` reservations, we ignore it, as we have skipped that case

msherman64 reviewed (approved):
LGTM in that I think it does the correct thing. I've left some comments as I think it could be made easier to understand, but that would also make it less flexible if we wanted to later add handling the case for fixing active or pending reservations too.

AnishReddyRavula committed:
rename method and filter out active reservations from DB
